pkg/manager/internal.go

```diff
@@ -582,12 +606,84 @@ func (cm *controllerManager) startLeaderElectionRunnables() {
 	for _, c := range cm.leaderElectionRunnables {
 		// Controllers block, but we want to return an error if any have an error starting.
 		// Write any Start errors to a channel so we can return them
-		cm.startRunnable(c)
+		cm.startRunnable(c, cm.internalStop)
 	}
 
 	cm.startedLeader = true
 }
 
+func (cm *controllerManager) startConditionalRunnables() {
```

Comment:
hi @kevindelgado , thanks for this PR. It seems to be its a bit chaotic and involves a set of semi-related changes, like the removal of informers which is IIRC not something we have agreed on and definitely something for a distinct PR.

As for the `conditionanally runnables` idea, a much simpler approach would be to just check whatever you want to check before adding them to the `manager`:
```
if clusterHasMyCRD(){
    controller.New()
}
```

This also has the advantage that you can check anything, not whatever happens to be implemented for this in the `manager`.

Reply:
Thanks for the feedback! 

We’re definitely open to better ways of doing this, but we are currently doing what you suggest. Outside of controller-runtime we conditionally start the manager based on the existence of the CRD. A couple issues we see are:

1. Increasing complexity. Specifically, when multiple controllers are abstracted to appear as a single controller or when the manager is running multiple controllers that watch a mixture of CRDs that will always be installed on the cluster, but also CRDs that are only sometimes installed. 
2. This approach still isn’t great for fluidly handling CRD install/uninstall/reinstall (rather than just waiting for install before starting), especially when latency becomes important immediately upon CRD installation.

/cc @DirectXMan12  in case they can better explain our desire to push this functionality further down the stack into c-r.

Reply:
This sounds to me like you should probably build a metacontroller that watches crds and then starts and stops the actual controllers. 

If you continue to think this requires changes in controller-runtime, the first step would be a design doc that describes the use-case, why this is problematic with how controller-runtime works currently, what changes would be required and possible alternatives so we can properly discuss this. We have the /designs folder for this.

Reply:
Sure thing, we’re working to create a more compelling story for this and will have a design doc up if/when that is the case. 

Sorry for the lack of context here, when learning the codebase it was easier to express ideas through concrete code than in prose initially. 

While the code has been helpful to steer discussions and now make it easier to formalize into a design doc, I definitely understand that it appears chaotic and disorganized externally. Feel free to close if you want to reduce clutter in the repo, I can always submit a new one if necessary.

Reply:
I think we can probably move a lot of this into a separate helper runnable that manages its own sub-runnables and managed the start/stop of things.  We'd still need a way to start/stop caches though, and possible some other minor tweaks.

I think the value of having this *somewhere* in controller-runtime is that if folks want to do something like this, they don't have to re-implement from scratch, since there are some corner cases.  I've seen it pop up a few times from various users on our side, and at the very least I think it'd be useful to have some basic hooks for better being able to start/stop controllers in CR, especially for multi-cluster usecases, where you don't necessarily want to/can't degrade performance for all clusters just to deal with a change in one. It also provides a motivating minimal usecase for starting/stopping caches.

Reply:
I think one of the things to keep in mind here is that if we go with the approach of "oh, you should just build it as a meta-controller", we need to make sure the hooks are there to do it cleanly, and we need to make sure those hooks are documented as such so that they don't get removed later on down the road.

Reply:
> At the very least I think it'd be useful to have some basic hooks for better being able to start/stop controllers in CR, especially for multi-cluster usecases, where you don't necessarily want to/can't degrade performance for all clusters just to deal with a change in one. It also provides a motivating minimal usecase for starting/stopping caches

Sure thing, but today we don't have any official story at all for multi-cluster _without the assumption that clusters might come and go_. I am not objecting to the basic idea (although I still think that if this is about being fast as mentioned above, a metacontroller would be the best), I just think that this requires a lot of groundwork first.